src/bn256/assembly.rs

```diff
@@ -1,5 +1,320 @@
 macro_rules! assembly_field {
-    ($field:ident, $modulus:ident, $inv:ident) => {
+    (
+        $field:ident,
+        $modulus:ident,
+        $inv:ident,
+        $modulus_str:ident,
+        $two_inv:ident,
+        $root_of_unity_inv:ident,
+        $delta:ident,
+        $zeta:ident
+    ) => {
+            use std::arch::asm;
+
+            impl $field {
+                /// Returns zero, the additive identity.
+                #[inline]
+                pub const fn zero() -> $field {
+                    $field([0, 0, 0, 0])
+                }
+
+                /// Returns one, the multiplicative identity.
+                #[inline]
+                pub const fn one() -> $field {
+                    R
+                }
+
+                fn from_u512(limbs: [u64; 8]) -> $field {
+                    // We reduce an arbitrary 512-bit number by decomposing it into two 256-bit digits
+                    // with the higher bits multiplied by 2^256. Thus, we perform two reductions
+                    //
+                    // 1. the lower bits are multiplied by R^2, as normal
+                    // 2. the upper bits are multiplied by R^2 * 2^256 = R^3
+                    //
+                    // and computing their sum in the field. It remains to see that arbitrary 256-bit
+                    // numbers can be placed into Montgomery form safely using the reduction. The
+                    // reduction works so long as the product is less than R=2^256 multiplied by
+                    // the modulus. This holds because for any `c` smaller than the modulus, we have
+                    // that (2^256 - 1)*c is an acceptable product for the reduction. Therefore, the
+                    // reduction always works so long as `c` is in the field; in this case it is either the
+                    // constant `R2` or `R3`.
+                    let d0 = $field([limbs[0], limbs[1], limbs[2], limbs[3]]);
+                    let d1 = $field([limbs[4], limbs[5], limbs[6], limbs[7]]);
+                    // Convert to Montgomery form
+                    d0 * R2 + d1 * R3
+                }
+
+                /// Converts from an integer represented in little endian
+                /// into its (congruent) `$field` representation.
+                pub const fn from_raw(val: [u64; 4]) -> Self {
```

Comment:
The non-asm version of this function is implemented as:
```
                (&$field(val)).mul(&R2)
```

Can this be applied here?

Reply:
Unfortunately no, because with `asm` the `mul` is no longer `const`, so we need to pick underlying logic manually to keep it `const` (it should be fine cause this function is usually used to initialize parameters).